build: Add support for GCC UBSan

DanielJerolm opened the pull request:
This PR adds support for building the unittest executable with GCC's Undefined Behaviour Sanitizer.

The approach is:
- Add a toolchain file which sets the required settings for UBSan globally for all build artifacts
- Scripts have been added to configure a third build folder (./build_unittest-ubsan), perform build actions in the folder, and for execution of the unittest executable.

A third build-folder has been added, because CMake does not reconfigure the build-folder properly if it is used for both the existing unittest build and the new build incorporating UBSan.

Open points:
- Provide a second badge? (e.g. "UB Sanitizer passed")
- ...or add/update to the existing build-action for unittests?

DanielJerolm committed:
Enhance doc and output of build scripts

DanielJerolm committed:
Add scripts to configure and execute a unittest build with UBSan

DanielJerolm committed:
Update README.mp

DanielJerolm committed:
Update history

Wishmaster-de commented on scripts/execute_unittests-ubsan.sh:
I would suggest to rename this script to execute_unittests_ubsan.sh.
Reason: All other execute scripts have an underscore in its name.

DanielJerolm replied:
Would you also suggest to rename the build-folder, too?
```
./build_productuive
./build_unittest
./build_unittest-ubsan
```
->
```
./build_productuive
./build_unittest
./build_unittest_ubsan
```

DanielJerolm replied:
And what about the other scripts: I used the underscore to separate words and the minus to separate configuration variants like "debug" and "release":
```
cmake_config_auto-productive-release.sh
cmake_config_auto-unittest-debug.sh
cmake_config_auto-unittest-release.sh
cmake_config_nativelinux-productive-release.sh
cmake_config_nativelinux-unittest-debug-notfc.sh
cmake_config_nativelinux-unittest-debug.sh
cmake_config_nativelinux-unittest-release-notfc.sh
cmake_config_nativelinux-unittest-release.sh
cmake_config_nativelinux-unittest-ubsan.sh
```
Style is not consistent I admit.

Wishmaster-de replied:
> And what about the other scripts: I used the underscore to separate words and the minus to separate configuration variants like "debug" and "release":
> 
> ```
> cmake_config_auto-productive-release.sh
> cmake_config_auto-unittest-debug.sh
> cmake_config_auto-unittest-release.sh
> cmake_config_nativelinux-productive-release.sh
> cmake_config_nativelinux-unittest-debug-notfc.sh
> cmake_config_nativelinux-unittest-debug.sh
> cmake_config_nativelinux-unittest-release-notfc.sh
> cmake_config_nativelinux-unittest-release.sh
> cmake_config_nativelinux-unittest-ubsan.sh
> ```
> 
> Style is not consistent I admit.

I think a consistent naming would be good, but it is not a major issue.

Wishmaster-de commented:
Open points:
Provide a second badge? (e.g. "UB Sanitizer passed")  ...or add/update to the existing build-action for unittests?

Currently it is only possible to have one badge for one workflow file. Maybe the developers will add the feature to have one badge per job in a workflow. That would be nice.
So i if we would like to have a badge for undefined behaviour sanitizer we have to add a new workflow.

DanielJerolm committed:
Merge branch 'develop' into feature/support-gcc-ubsan

DanielJerolm commented:
Is there any advantage in having two workflows?
I would tend to use only one, but then build the unittests always with UBsan.

Wishmaster-de commented:
> Is there any advantage in having two workflows? I would tend to use only one, but then build the unit tests always with UBsan.

First I thought there would be something. But to have two workflows has only the advantage that the "normal" unit test build would have no extra instrumentation code like the undefined behavior build.  But the unit tests should work with and  without instrumentation code, maybe with instrumentation code a little bit slower?

I agree with you.

DanielJerolm committed:
Revise build configurations and scripts

There are three build-folders for unittests now:
- build_unittest for daily use

- build_unittest-buildsrc for the buildserver;
Applies memcheck and UBSan

- build_unittest-notfc for local execution of testcases
that cannot make use of TFC and that depend on
machine load

DanielJerolm commented:
I have revised the required build-configurations and cleaned them up.

There are three flavors of unittests now, each having its own build-folder. For each build-folder there are configure, build, and execute scripts.

- build_unittest: For everyday use in develoment.
- build_unittest-buildsrv: For the build server. Applies UBSan and memcheck. These tests may also be executed locally, but build-time and run-time are about 56-58% larger on my machine.
- build_unittest-notfc: Builds unittest without presence of TFC. This enables a few tests for the OSAL that require absence of TFC. This is used on rare occasions and only locally.

DanielJerolm committed:
Add valgrind installation to build-test.yaml

DanielJerolm committed:
Enhance error checks

Wishmaster-de commented on scripts/execute_unittests-buildsrv.sh:
```diff
@@ -0,0 +1,44 @@
+elif [ $# -eq 1 ]; then
+
+  cd ../build_unittest-buildsrv/output
+  valgrind --tool=memcheck --leak-check=full --show-leak-kinds=definite,indirect,possible ./unittests --gtest_filter=$1
```
Maybe you should check if valgrind is installed  and if not print something useful for the user. But, this script is intended to be used for the buildserver where the owner is in behalf for installation of prerequisites this is maybe not useful.

DanielJerolm replied:
Added valgrind to list of host system requirements in README.md

Wishmaster-de reviewed (changes requested):
- Please update the README.txt e.G. cmake_config_nativelinux-unittest-debug.sh no longer exists
- Maybe the execute scripts should check if the output folder is available. If a user calls a Run Task for a different configuration the output folder isn't available.

DanielJerolm committed:
Update README.md and other documentation

DanielJerolm committed:
Harmonize arguments passed to execute-scripts

DanielJerolm committed:
Execute scripts check presence of executable

Wishmaster-de commented:
I tried to execute the unittests with notfc. I did it on Ubuntu Noble.
I am getting some failed tests:
`[  FAILED  ] 11 tests, listed below:
[  FAILED  ] gpcc_execution_async_DeferredWorkQueue_/IWorkQueue_Tests2F/0.WaitUntilCurrentWorkPackageHasBeenExecuted, where TypeParam = gpcc::execution::async::DeferredWorkQueue
[  FAILED  ] gpcc_execution_async_DeferredWorkQueue_/IWorkQueue_Tests2F/0.FlushNonDeferredWorkPackages, where TypeParam = gpcc::execution::async::DeferredWorkQueue
[  FAILED  ] gpcc_execution_async_DeferredWorkQueue_TestsF.Deferred_WaitUntilCurrentWorkPackageHasBeenExecuted
[  FAILED  ] gpcc_execution_async_DeferredWorkQueue_TestsF.InsertionWithShorterDelay
[  FAILED  ] gpcc_execution_async_DeferredWorkQueue_TestsF.DeferredWPbecomesRunnableDuringWPExec
[  FAILED  ] gpcc_execution_async_WorkQueue_/IWorkQueue_Tests2F/0.WaitUntilCurrentWorkPackageHasBeenExecuted, where TypeParam = gpcc::execution::async::WorkQueue
[  FAILED  ] gpcc_execution_async_WorkQueue_/IWorkQueue_Tests2F/0.FlushNonDeferredWorkPackages, where TypeParam = gpcc::execution::async::WorkQueue
[  FAILED  ] gpcc_osal_Thread_TestsF.Start_Policy_SP_Idle
[  FAILED  ] gpcc_osal_Thread_TestsF.Start_Policy_SP_Batch
[  FAILED  ] gpcc_osal_Thread_TestsF.Start_Policy_SP_Fifo
[  FAILED  ] gpcc_osal_Thread_TestsF.Start_Policy_SP_RR`

DanielJerolm committed:
Do not run TFC-based tests if TFC is not available.

DanielJerolm commented:
I checked the failed testcases.
All failures are plausible:
- Some failed because the configuration option for excluding TFC based tests was not set. I fixed that in the `unittest-notfc` build configuration
- Some failes due to missing user permissions (setting RT-scheduling policies). I enhanced the unittest configuration doc and the messages displayed by the scripts for building and executing the "special" tests
